source/common/ssl/tls_certificate_config_impl.cc

```diff
@@ -22,9 +23,37 @@ TlsCertificateConfigImpl::TlsCertificateConfigImpl(
                             .value_or(private_key_.empty() ? EMPTY_STRING : INLINE_STRING)),
       password_(Config::DataSource::read(config.password(), true, api)),
       password_path_(Config::DataSource::getPath(config.password())
-                         .value_or(password_.empty() ? EMPTY_STRING : INLINE_STRING)) {
+                         .value_or(password_.empty() ? EMPTY_STRING : INLINE_STRING)) {}
 
-  if (certificate_chain_.empty() || private_key_.empty()) {
+TlsCertificateConfigImpl::TlsCertificateConfigImpl(
+    const envoy::api::v2::auth::TlsCertificate& config, Api::Api& api,
+    bool expect_private_key_method)
```

Comment:
The reason for the private constructor was an attempt to avoid copying the initialization code into the two public constructors. When the constructor is called from SDS initialization, the error handling logic goes bit differently, because there is no possibility to set the private key method (the `expect_private_key_method` value will be false) and the factory context isn't available. The separate constructor was needed because I wanted to have factory_context as a reference.

But you're right that the logic is complex. I'll look if I can make it simpler.